portfolio/src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java

```diff
@@ -0,0 +1,189 @@
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * When the user submits the form, Blobstore processes the file upload and then forwards the request
+ * to this servlet. This servlet can then analyze the image using the Vision API.
+ */
+@WebServlet("/image-analysis")
+public class ImageAnalysisServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    PrintWriter out = response.getWriter();
+
+    // Get the message entered by the user.
+    String message = request.getParameter("message");
+
+    // Get the BlobKey that points to the image uploaded by the user.
+    BlobKey blobKey = getBlobKey(request, "image");
+
+    // User didn't upload a file, so render an error message.
+    if (blobKey == null) {
+      out.println("Please upload an image file.");
+      return;
+    }
+
+    // Get the URL of the image that the user uploaded.
+    String imageUrl = getUploadedFileUrl(blobKey);
+
+    // Get the labels of the image that the user uploaded.
+    byte[] blobBytes = getBlobBytes(blobKey);
+    List<EntityAnnotation> imageLabels = getImageLabels(blobBytes);
+
+    // Output some HTML that shows the data the user entered.
+    // A real codebase would probably store these in Datastore.
+    response.setContentType("text/html");
+    out.println("<p>Here's the image you uploaded:</p>");
+    out.println("<a href=\"" + imageUrl + "\">");
+    out.println("<img src=\"" + imageUrl + "\" />");
+    out.println("</a>");
+    out.println("<p>Here are the labels we extracted:</p>");
+    out.println("<ul>");
+    for (EntityAnnotation label : imageLabels) {
+      out.println("<li>" + label.getDescription() + " " + label.getScore());
+    }
+    out.println("</ul>");
+  }
+
+  /**
+   * Returns the BlobKey that points to the file uploaded by the user, or null if the user didn't
+   * upload a file.
+   */
+  private BlobKey getBlobKey(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+
+    // User submitted form without selecting a file, so we can't get a BlobKey. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so the BlobKey is empty. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    return blobKey;
+  }
+
+  /**
+   * Blobstore stores files as binary data. This function retrieves the binary data stored at the
+   * BlobKey parameter.
```

Comment:
It would be good to update the comment to note an assumptions for this function.

For example: do you require the blobkey exists?  What happens if it doesn't exist?